fix(parse): Implement amended "Rule 2"

gwhitney opened the pull request:
As per the discussion in #2370, the amended "Rule 2" is
  "when having a division followed by an implicit multiplication, the
   division gets higher precedence over the implicit multiplication when
   (a) the numerator is a constant with optionally a
       prefix operator (-, +, ~), and
   (b) the denominator is a constant."
  This commit implements that behavior and adds tests for it.
  Resolves #2370.

josdejong commented:
Looks very good. It's nice to see the special rule contained in a single `isRule2Node` function (and not spread out all over the place).

Before merging, I would love to see if we can solve the FIXME too (see https://github.com/josdejong/mathjs/issues/1431#issuecomment-1057820726 and my reply on that)

josdejong commented on test/unit-tests/expression/parse.test.js:
```diff
@@ -1336,11 +1336,16 @@ describe('parse', function () {
 
     it('should follow precedence rules for implicit multiplication and division', function () {
       assert.strictEqual(parseAndStringifyWithParens('2 / 3 x'), '(2 / 3) x')
+      assert.strictEqual(parseAndStringifyWithParens('-2/3x'), '((-2) / 3) x')
```
@gwhitney are there also tests in place to check how non-constant nodes in numerator/denominator are handled now? Like:  `-(2+3)/3x`, `-2/(3+4)x`, etc. I.e. the cases you listed in https://github.com/josdejong/mathjs/issues/2370#issuecomment-1031077094

gwhitney replied:
OK, have added any more test cases in both parse.test.js and OperatorNode.test.js.

gwhitney commented:
> I would love to see if we can solve the FIXME too (see [#1431 (comment)](https://github.com/josdejong/mathjs/issues/1431#issuecomment-1057820726) and my reply on that)

Ah, OK. I will incorporate a revised version of #2403 into this PR so that it will simultaneously resolve #1431.

josdejong commented:
Thanks!

gwhitney commented:
OK, I have done this, but you may not be too happy. This was in fact the proverbial can of worms. The more tests I added, the more tricky cases popped up, and the section that handles special parenthesization rules in `OperatorNode.toString()` has become very large and complicated as a result. But I believe that at least one of the cases in the new test suite will fail if any clause of the logic is eliminated. It would be nice if there were some more systematic approach to determining the necessary parentheses in the face of implicit multiplication, that would lead to cleaner code, but I was unable to come up with one. I am open to any thoughts you might have.

That said, I wouldn't recommend not merging any version of this branch. The more extensive test suite revealed many inconsistencies between .toString() and parsing in develop, so avoiding this because the code is ugly won't make the existing bugs go away... Maybe this can be a stopgap until someone who can devote more thought to the structure of the problem can find a cleaner solution to how to parenthesize a parse tree.

Note that some completely unrelated bugs were shaken out and fixed, such as toTex on a multi-argument implicit multiplication not respecting the 'implicit' flag.

Let me know what else you'd like me to do here to move this toward merging; it would be nice to get this parsing and formatting stuff for implicit multiplication by fractions behind us.

I will say, though, it would be good if you could scan through the implicit multiplication tests in both parse.test.js and OperatorNode.test.js to check that the prescribed outcomes really do match what's truly intended for mathjs behavior. I think I got them all right, but I did start to get cross-eyed with all of the fractions and multiplications and parentheses.

Thanks for working with this.

josdejong commented:
Thanks for the updates. I'll have a thorough look at it today or tomorrow. The logic needed in `calculateNecessaryParentheses` looks quite complex indeed at first sight 😅.

I've marked this PR as a breaking change to be scheduled for `v11`

josdejong commented:
Wooow that's a huge amount of work you've put in there, thanks! I hadn't expected this to get this complex 😮 

Thanks for spending so much time on the unit tests, that really gives confidence. I do like your refactoring in `OperatorNode.test.js`, that makes the tests more compact to read/write, and the `ex` and `exhs` utility functions are really handy to test a lot of scenarios more easily. I've gone though the test cases and they look good, this is exactly the behavior what we wanted 👌. Impressive job again!

Seen the complexity needed to get `toString()` right, I have two more thoughts:

1. Maybe it is acceptable to be a bit too conservative and have some cases where parenthesis are outputted when not strictly necessary if that would simplify the logic needed in `toString()`. Most important goals are: when parsing a stringified expression again we want to get the same mathematical expression again, and second: we want to see as little parentheses as needed in the stringified output with "auto" mode. We could be a bit pragmatic on this second goal, keep the code simple, and just stay on the "safe side" in case of doubt?
2. Would it be possible to "go dumb" here in `toString()` when dealing with implicit multiplication? We could just try out stringifying and expression without parenthesis, then parse the string again and see whether the parsed result is identical to the original node tree. If not, we apparently have to add parenthesis. This is not very efficient but could be acceptable: it's only applied in specific cases, when dealing with implicit multiplications.

gwhitney commented:
OK, I think you will find this much better. After some thought, the key to simplifying the parenthesis computation was to provide for a distinct precedence level of implicit multiplication, and actually return that precedence level from operators.js -- then there is only one "special case" in OperatorNode:calculateNecessaryParentheses, for subexpressions of an implicit multiplication that start with a constant.  (There's another small price to pay, which is that the "implicit" flag must always be passed in to `getPrecedence()` but that seemed like a small cost for making the parenthesization code actually readable.)

Let me know what you think of the PR now. I am unclear on why automated tests do not seem to be running here in GitHub anymore -- is it because of now being on top of v11? In any case, they all passed on my computer, but that doesn't run the browser tests.

josdejong commented:
grr. I've reverted the change from `options && options.implicit` to `options?.implicit` again. It breaks the unit tests on Nodejs 12 where optional chaining doesn't yet exist. Let's be lazy and wait 1.5 months, when the end of life of Node.js 12 is reached and we can drop support for it.

gwhitney commented:
> 1. The unit test "should return the precedence of a node" was failing. I changed

Yes, I expected to have to make that change. Maybe I didn't see it in my tests because not all of the prior changes were in. I will make sure to work on top of V11 for anything further that goes on top of that. And yes, it went up by one to make room for the "implicit multiplication" precedence.

On `options?.implicit` -- I didn't use that because I had thought it was only TypeScript. Nice to know it is heading toward standard in JavaScript, too.
